src/network/networkd-ndisc.c

```diff
@@ -9,16 +9,86 @@
 #include "sd-ndisc.h"
 
 #include "missing_network.h"
+#include "parse-util.h"
+#include "random-util.h"
 #include "networkd-dhcp6.h"
 #include "networkd-manager.h"
 #include "networkd-ndisc.h"
 #include "networkd-route.h"
+#include "string-table.h"
+#include "string-util.h"
 #include "strv.h"
 
 #define NDISC_DNSSL_MAX 64U
 #define NDISC_RDNSS_MAX 64U
 #define NDISC_PREFIX_LFT_MIN 7200U
 
+#define DAD_CONFLICTS_IDGEN_RETRIES_RFC7217 3
+
+/* https://tools.ietf.org/html/rfc5453 */
+/* https://www.iana.org/assignments/ipv6-interface-ids/ipv6-interface-ids.xml */
+
+#define SUBNET_ROUTER_ANYCAST_ADDRESS_RFC4291 ((struct in6_addr) { .s6_addr = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 } })
+#define RESERVED_IPV6_INTERFACE_IDENTIFIERS_ADDRESS_RFC4291 ((struct in6_addr) { .s6_addr = { 0x02, 0x00, 0x5E, 0xFF, 0xFE } })
+#define RESERVED_SUBNET_ANYCAST_ADDRESSES_RFC4291 ((struct in6_addr) { .s6_addr = { 0xFD, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF } })
+
+#define NDISC_APP_ID SD_ID128_MAKE(13,ac,81,a7,d5,3f,49,78,92,79,5d,0c,29,3a,bc,7e)
+
+static bool ipv6_interface_identifiers(const struct in6_addr *addr) {
+
+        assert(addr);
+
+        if (!memcmp(addr, &SUBNET_ROUTER_ANYCAST_ADDRESS_RFC4291, 8))
+                return true;
+
+        if (!memcmp(addr, &RESERVED_IPV6_INTERFACE_IDENTIFIERS_ADDRESS_RFC4291, 5))
+                return true;
+
+        if (!memcmp(addr, &RESERVED_SUBNET_ANYCAST_ADDRESSES_RFC4291, 7))
+                return true;
+
+        return false;
+}
+
+static int ipv6_makestableprivate(Link *link, const struct in6_addr *prefix, int prefix_len, uint8_t *dad_counter, struct in6_addr *addr) {
+        char secret_key_buf[SD_ID128_STRING_MAX];
+        sd_id128_t secret_key;
+        struct siphash state;
+        uint64_t rid;
+        size_t len, l;
+        int r;
+
+        if (*dad_counter > DAD_CONFLICTS_IDGEN_RETRIES_RFC7217)
+                return log_error_errno(SYNTHETIC_ERRNO(EBUSY),
+                                       "Failed to generate prefix stable. Reached maximum DAD conflicts.");
+
+        /* According to rfc7217 section 5.1
+           RID = F(Prefix, Net_Iface, Network_ID, DAD_Counter, secret_key)
+        */
+        siphash24_init(&state, link->ndisc_hash_key);
+
+        siphash24_compress(&prefix, prefix_len, &state);
+        siphash24_compress(&link->ifname, strlen(link->ifname), &state);
+        siphash24_compress(&link->mac, sizeof(struct ether_addr), &state);
+        siphash24_compress(dad_counter, sizeof(uint8_t), &state);
+
+        r = sd_id128_get_machine_app_specific(NDISC_APP_ID, &secret_key);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get machine id for prefix stable private: %m");
```

Comment:
Error messages must be understandable for the user. Don't use terms which only make sense internally in the code.